README.md
# dropdown_plus

Simple and easy to use Dropdown in forms with search, keyboard navigation, offiline data source, remote data source and easy customization.

## Getting Started

Simple Text Dropdown.

![](https://github.com/crawlinknetworks/dropdown_plus/blob/master/screenshots/screen1.png?raw=true)
![](https://github.com/crawlinknetworks/dropdown_plus/blob/master/screenshots/screen6.png?raw=true)

```
TextDropdownFormField<String>(
    options: ["Male", "Female"],
    decoration: InputDecoration(
        border: OutlineInputBorder(),
        suffixIcon: Icon(Icons.arrow_drop_down),
        labelText: "Gender"),
    dropdownMaxHeight: 120,
),
```

## Install

##### packages.yaml
```
dropdown_plus: <lastest version>
```

## Customizable Example

<p align="left">
  <img src="https://github.com/crawlinknetworks/dropdown_plus/blob/master/screenshots/screen4.png?raw=true" width="300">
  <img src="https://github.com/Willian199/dropdown_plus/blob/master/screenshots/screen7.jpg?raw=true" width="300">
  <img src="https://github.com/Willian199/dropdown_plus/blob/master/screenshots/screen8.jpg?raw=true" width="300">
</p>

```

  String rolesJson = '''
    [
      {"name": "Super Admin", "desc": "Having full access rights", "role": 1},
      {"name": "Admin", "desc": "Having full access rights of a Organization", "role": 2},
      {"name": "Manager", "desc": "Having Magenent access rights of a Organization", "role": 3},
      {"name": "Technician", "desc": "Having Technician Support access rights", "role": 4},
      {"name": "Customer Support", "desc": "Having Customer Support access rights", "role": 5},
      {"name": "User", "desc": "Having End User access rights", "role": 6}
    ]
  ''';

// ...
// ...
 List<Role> roles = Role.rolesFromJson();


// ...
// ...

  DropdownFormField<Role>(
      dropdownColor: Theme.of(context).colorScheme.secondary,
      emptyText: 'Role not found',
      dropdownMaxHeight: 350,
      searchCursorColor: Colors.grey[800],
      searchInpuType: TextInputType.text,
      searchInitialValue: roles[5],
      autoFocus: false,
      searchTextStyle: TextStyle(
        color: Theme.of(context).colorScheme.secondaryContainer,
        fontSize: 18,
      ),
      controller: _controllerRole,
      decoration: InputDecoration(
        border: const UnderlineInputBorder(),
        suffixIcon: const Icon(Icons.arrow_drop_down),
        hintText: "Select a Role",
        errorBorder: const UnderlineInputBorder(),
        errorText: _errorMessage,
      ),
      emptyActionText: 'Create a new Role',
      onEmptyActionPressed: (String valueFromfield) async {
        setState(() {
          roles.add(Role(
              name: valueFromfield,
              description: 'Set a description for: $valueFromfield',
              id: roles.map((Role e) => e.id).reduce(max)));
        });
      },
      onSaved: (Role? str) {
        _showModal('Role ${str?.name} as been selected.');
      },
      onChanged: (Role? str) {
        _showModal('Role ${str?.name ?? ''} as been changed.');
      },
      validator: (Role? str) {
        if (_errorMessage == null && str == null) {
          return null;
        }
        _errorMessage =
            str != null && str.name == 'User' ? 'Oh noo' : null;

        return _errorMessage;
      },
      displayItemFn: (Role? item) => Text(
        (item?.name ?? ''),
        style: const TextStyle(fontSize: 16),
      ),
      //can make a http request
      findFn: (dynamic str) async => roles,
      selectedFn: (Role? item1, Role? item2) {
        if (item1 != null && item2 != null) {
          return item1.name == item2.name;
        }
        return false;
      },
      filterFn: (Role? item, String str) =>
          item?.name.toLowerCase().contains(str.toLowerCase()) ?? false,
      dropdownItemFn: (Role? item, int position, bool focused,
          bool selected, void Function() onTap) {
        return Column(
          children: [
            ListTile(
              title: Text(
                item?.name ?? '',
                style: TextStyle(
                  color: selected ? Colors.white : Colors.black,
                ),
              ),
              subtitle: Text(
                item?.description ?? '',
                style: TextStyle(
                  color: selected ? Colors.white : Colors.black,
                ),
              ),
              tileColor: Colors.transparent,
              onTap: onTap,
            ),
            const Divider(
              height: 3,
              color: Colors.white60,
            )
          ],
        );
      },
    )
  ),
```

## Options

```
  /// [filterFn] An optional function that takes a generic type `T` and a string and returns a boolean. <br>
  /// This function is used to filter the items in the dropdown list based on the search text entered by the user.
  final bool Function(T item, String str)? filterFn;

  /// Check item is selectd
  final bool Function(T? item1, T? item2)? selectedFn;

  /// [dropdownItemFn] A required function that takes a generic type `T` and returns a widget that represents a dropdown item for that value. <br>
  /// This function is used to render each item in the dropdown list.
  /// [item] = [dynamic value] List item to build dropdown Listtile
  /// [lasSelectedItem] = [null | dynamic value] last selected item, it gives user chance to highlight selected item
  /// [position] = [0,1,2...] Index of the list item
  /// [focused] = [true | false] is the item if focused, it gives user chance to highlight focused item
  /// [onTap] = [Function] *important! just assign this function to Listtile.onTap  = onTap, incase you missed this,
  /// the click event if the dropdown item will not work.
  ///
  final Widget Function(
    T item,
    int position,
    bool focused,
    bool selected,
    void Function() onTap,
  ) dropdownItemFn;

  /// An optional function that takes a generic type T and is called when the selected value is changed.
  final void Function(T? item)? onChanged;

  /// An optional function that takes a generic type T and is called when the form is saved.
  final void Function(T?)? onSaved;

  /// An optional function that takes a generic type T and returns a string. This function is used to validate the selected value.
  final String? Function(T? item)? validator;

  /// this functon triggers on click of emptyAction button
  final Future<void> Function(String)? onEmptyActionPressed;

  ///An optional boolean that determines whether the dropdown field should be focused automatically when the widget is loaded.
  final bool autoFocus;
  final DropdownEditingController<T>? controller;

  /// An optional decoration for the text field that is used to enter search text.
  final InputDecoration? decoration;

  /// [displayItemFn] A required function that takes a generic type `T` and returns a string that represents the display text for that value. <br>
  /// This function is used to display the selected value in the dropdown field.
  final Widget Function(T? item) displayItemFn;

  /// An optional color for the dropdown list background.
  final Color? dropdownColor;

  /// Max height of the dropdown overlay, Default: 240
  final double? dropdownMaxHeight;

  /// Give action text if you want handle the empty search.
  final String emptyActionText;

  /// Message to display if the search dows not match with any item, Default : "No matching found!"
  final String emptyText;

  /// [findFn] A required function that takes a string and returns a Future that resolves to a list of generic type. <br>
  /// This function is used to retrieve the list of items that match the search text entered by the user.
  /// Return list of items what need to list for dropdown.
  /// The list may be offline, or remote data from server.
  final Future<List<T>> Function(String str) findFn;

  /// Cursor color of the search box text
  final Color? searchCursorColor;

  /// An optional initial value for the search text field.
  final T? searchInitialValue;

  /// TextInputType of the search box text
  final TextInputType searchInpuType;

  /// An optional style for the search text field.
  final TextStyle? searchTextStyle;

```


CHANGELOG.md
## 1.0.0

* Code refactor and implemented some changes
* Added possibility to clear values
* Fixed issues:
    [#30](https://github.com/crawlinknetworks/dropdown_plus/issues/30)
    [#22](https://github.com/crawlinknetworks/dropdown_plus/issues/22)
    [#20](https://github.com/crawlinknetworks/dropdown_plus/issues/20)
    [#18](https://github.com/crawlinknetworks/dropdown_plus/issues/18)
    [#16](https://github.com/crawlinknetworks/dropdown_plus/issues/16)
    [#7](https://github.com/crawlinknetworks/dropdown_plus/issues/7)
    [#4](https://github.com/crawlinknetworks/dropdown_plus/issues/4)

## 0.0.8

* Fixed onSave issue.

## 0.0.8

* DropdownEditingController added.

## 0.0.7

* Bug Fixes.

## 0.0.6

* selectedFn added.

## 0.0.5

* Widget name Change to DropdownFormField.

## 0.0.4

* Bug fix value change.

## 0.0.3

* Documentation updated.

## 0.0.2

* Bug fix release.

## 0.0.1

* First release.


lib/src/dropdown.dart
import 'dart:async';

import 'package:dropdown_plus/src/dropdown_controller.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// Create a dropdown form field
class DropdownFormField<T> extends StatefulWidget {
  const DropdownFormField({
    required this.dropdownItemFn,
    required this.displayItemFn,
    required this.findFn,
    Key? key,
    this.filterFn,
    this.autoFocus = false,
    this.controller,
    this.validator,
    this.decoration,
    this.dropdownColor,
    this.onChanged,
    this.onSaved,
    this.dropdownMaxHeight,
    this.searchTextStyle,
    this.emptyText = "No matching found!",
    this.emptyActionText = 'Create new',
    this.onEmptyActionPressed,
    this.selectedFn,
    this.searchCursorColor,
    this.searchInitialValue,
    this.searchInpuType = TextInputType.emailAddress,
  }) : super(key: key);

  /// [filterFn] An optional function that takes a generic type `T` and a string and returns a boolean. <br>
  /// This function is used to filter the items in the dropdown list based on the search text entered by the user.
  final bool Function(T item, String str)? filterFn;

  /// Check item is selectd
  final bool Function(T? item1, T? item2)? selectedFn;

  /// [dropdownItemFn] A required function that takes a generic type `T` and returns a widget that represents a dropdown item for that value. <br>
  /// This function is used to render each item in the dropdown list.
  /// [item] = [dynamic value] List item to build dropdown Listtile
  /// [lasSelectedItem] = [null | dynamic value] last selected item, it gives user chance to highlight selected item
  /// [position] = [0,1,2...] Index of the list item
  /// [focused] = [true | false] is the item if focused, it gives user chance to highlight focused item
  /// [onTap] = [Function] *important! just assign this function to Listtile.onTap  = onTap, incase you missed this,
  /// the click event if the dropdown item will not work.
  ///
  final Widget Function(
    T item,
    int position,
    bool focused,
    bool selected,
    void Function() onTap,
  ) dropdownItemFn;

  /// An optional function that takes a generic type T and is called when the selected value is changed.
  final void Function(T? item)? onChanged;

  /// An optional function that takes a generic type T and is called when the form is saved.
  final void Function(T?)? onSaved;

  /// An optional function that takes a generic type T and returns a string. This function is used to validate the selected value.
  final String? Function(T? item)? validator;

  /// this functon triggers on click of emptyAction button
  final Future<void> Function(String)? onEmptyActionPressed;

  ///An optional boolean that determines whether the dropdown field should be focused automatically when the widget is loaded.
  final bool autoFocus;
  final DropdownEditingController<T>? controller;

  /// An optional decoration for the text field that is used to enter search text.
  final InputDecoration? decoration;

  /// [displayItemFn] A required function that takes a generic type `T` and returns a string that represents the display text for that value. <br>
  /// This function is used to display the selected value in the dropdown field.
  final Widget Function(T? item) displayItemFn;

  /// An optional color for the dropdown list background.
  final Color? dropdownColor;

  /// Max height of the dropdown overlay, Default: 240
  final double? dropdownMaxHeight;

  /// Give action text if you want handle the empty search.
  final String emptyActionText;

  /// Message to display if the search dows not match with any item, Default : "No matching found!"
  final String emptyText;

  /// [findFn] A required function that takes a string and returns a Future that resolves to a list of generic type. <br>
  /// This function is used to retrieve the list of items that match the search text entered by the user.
  /// Return list of items what need to list for dropdown.
  /// The list may be offline, or remote data from server.
  final Future<List<T>> Function(String str) findFn;

  /// Cursor color of the search box text
  final Color? searchCursorColor;

  /// An optional initial value for the search text field.
  final T? searchInitialValue;

  /// TextInputType of the search box text
  final TextInputType searchInpuType;

  /// An optional style for the search text field.
  final TextStyle? searchTextStyle;

  @override
  DropdownFormFieldState<T> createState() => DropdownFormFieldState<T>();
}

class DropdownFormFieldState<T> extends State<DropdownFormField<T>>
    with SingleTickerProviderStateMixin {
  DropdownFormFieldState() : super();

  final DropdownEditingController<T>? _controller =
      DropdownEditingController<T>();

  Timer? _debounce;
  Widget? _displayItem;
  bool _isFocused = false;
  String? _lastSearchString;
  final LayerLink _layerLink = LayerLink();
  int _listItemFocusedPosition = 0;
  final ValueNotifier<List<T>> _listItemsValueNotifier =
      ValueNotifier<List<T>>(<T>[]);

  List<T>? _options;
  late OverlayEntry _overlayBackdropEntry;
  OverlayEntry? _overlayEntry;
  final FocusNode _searchFocusNode = FocusNode();
  final TextEditingController _searchTextController = TextEditingController();
  T? _selectedItem;
  final FocusNode _widgetFocusNode = FocusNode();

  @override
  void dispose() {
    super.dispose();
    _debounce?.cancel();
    _searchTextController.dispose();
    _widgetFocusNode.dispose();
  }

  @override
  void initState() {
    super.initState();

    if (widget.autoFocus) {
      _widgetFocusNode.requestFocus();
    }
    _selectedItem = _effectiveController!.value;

    _searchFocusNode.addListener(() {
      if (!_searchFocusNode.hasFocus && _overlayEntry != null) {
        _removeOverlay();
      }
    });

    _effectiveController!.addListener(() {
      if (_effectiveController!.value == null) {
        _clearValue();
      }
    });
  }

  final bool Function(T?, T?) _selectedFn =
      (T? item1, T? item2) => item1 == item2;

  bool get _isEmpty => _selectedItem == null;

  DropdownEditingController<T>? get _effectiveController =>
      widget.controller ?? _controller;

  OverlayEntry _createOverlayEntry() {
    final RenderBox renderObject = context.findRenderObject() as RenderBox;
    // print(renderObject);
    final Size size = renderObject.size;

    OverlayEntry overlay = OverlayEntry(
      builder: (BuildContext context) {
        return Positioned(
          width: size.width,
          child: CompositedTransformFollower(
            link: this._layerLink,
            showWhenUnlinked: false,
            offset: Offset(0.0, size.height + 3.0),
            child: Material(
              elevation: 10.0,
              color: Colors.transparent,
              shadowColor: Colors.transparent,
              child: Container(
                decoration: BoxDecoration(
                  color: widget.dropdownColor ?? Theme.of(context).primaryColor,
                  borderRadius: const BorderRadius.all(
                    Radius.circular(20),
                  ),
                ),
                constraints:
                    BoxConstraints(maxHeight: widget.dropdownMaxHeight ?? 240),
                child: ValueListenableBuilder<List<T>>(
                  valueListenable: _listItemsValueNotifier,
                  builder:
                      (BuildContext context, List<T> items, Widget? child) {
                    return _options?.isNotEmpty ?? false
                        ? ListView.builder(
                            shrinkWrap: true,
                            padding: EdgeInsets.zero,
                            itemCount: _options!.length,
                            itemBuilder: (BuildContext context, int position) {
                              T item = _options![position];

                              void Function() onTap = () {
                                _listItemFocusedPosition = position;
                                _searchTextController.value = TextEditingValue(
                                  text: item.toString(),
                                  selection: TextSelection.collapsed(
                                    offset: item.toString().length,
                                  ),
                                );
                                _removeOverlay();
                                _setValue();
                              };

                              Widget listTile = widget.dropdownItemFn(
                                item,
                                position,
                                position == _listItemFocusedPosition,
                                (widget.selectedFn ?? _selectedFn)(
                                    _selectedItem, item),
                                onTap,
                              );

                              return listTile;
                            })
                        : Container(
                            padding: const EdgeInsets.all(16),
                            height:
                                widget.onEmptyActionPressed == null ? 80 : 120,
                            child: Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              children: [
                                Text(
                                  widget.emptyText,
                                  style: TextStyle(
                                    color: Theme.of(context).primaryColor,
                                  ),
                                ),
                                if (widget.onEmptyActionPressed != null)
                                  ElevatedButton(
                                    onPressed: () async {
                                      await widget.onEmptyActionPressed!(
                                          _searchTextController.value.text);
                                      _search(_searchTextController.value.text);
                                    },
                                    child: Text(widget.emptyActionText),
                                  ),
                              ],
                            ),
                          );
                  },
                ),
              ),
            ),
          ),
        );
      },
    );

    return overlay;
  }

  OverlayEntry _createBackdropOverlay() {
    return OverlayEntry(
      builder: (BuildContext context) => Positioned(
        left: 0,
        top: 0,
        height: MediaQuery.of(context).size.height,
        width: MediaQuery.of(context).size.width,
        child: GestureDetector(
          onTap: () {
            _removeOverlay();
          },
        ),
      ),
    );
  }

  void _addOverlay() {
    if (_overlayEntry == null) {
      unawaited(_search(""));
      _overlayBackdropEntry = _createBackdropOverlay();
      _overlayEntry = _createOverlayEntry();
      if (_overlayEntry != null) {
        // Overlay.of(context)!.insert(_overlayEntry!);
        Overlay.of(context)
            .insertAll(<OverlayEntry>[_overlayBackdropEntry, _overlayEntry!]);
        setState(() {
          _searchFocusNode.requestFocus();
        });
      }
    }
  }

  /// Dettach overlay from the dropdown widget
  void _removeOverlay() {
    if (_overlayEntry != null) {
      _overlayBackdropEntry.remove();
      _overlayEntry!.remove();
      _overlayEntry = null;
      _searchTextController.value = TextEditingValue.empty;
      setState(() {});
    }
  }

  void _toggleOverlay() {
    if (_overlayEntry == null) {
      _addOverlay();
    } else {
      _removeOverlay();
    }
  }

  void _onTextChanged(String? str) {
    if (_debounce?.isActive ?? false) {
      _debounce!.cancel();
    }
    _debounce = Timer(const Duration(milliseconds: 300), () {
      if (_lastSearchString != str) {
        _lastSearchString = str;
        unawaited(_search(str ?? ""));
      }
    });
  }

  KeyEventResult _onKeyPressed(RawKeyEvent event) {
    if (event.isKeyPressed(LogicalKeyboardKey.enter)) {
      if (_searchFocusNode.hasFocus) {
        _toggleOverlay();
      } else {
        _toggleOverlay();
      }
      return KeyEventResult.ignored;
    } else if (event.isKeyPressed(LogicalKeyboardKey.escape)) {
      _removeOverlay();
      return KeyEventResult.handled;
    } else if (event.isKeyPressed(LogicalKeyboardKey.arrowDown)) {
      int v = _listItemFocusedPosition;
      v++;
      if (v >= _options!.length) {
        v = 0;
      }
      _listItemFocusedPosition = v;
      _listItemsValueNotifier.value = List<T>.from(_options ?? <T>[]);
      return KeyEventResult.handled;
    } else if (event.isKeyPressed(LogicalKeyboardKey.arrowUp)) {
      int v = _listItemFocusedPosition;
      v--;
      if (v < 0) {
        v = _options!.length - 1;
      }
      _listItemFocusedPosition = v;
      _listItemsValueNotifier.value = List<T>.from(_options ?? <T>[]);
      return KeyEventResult.handled;
    }
    return KeyEventResult.ignored;
  }

  Future<void> _search(String str) async {
    List<T> items = await widget.findFn(str);

    if (str.isNotEmpty && widget.filterFn != null) {
      items = items.where((T item) => widget.filterFn!(item, str)).toList();
    }

    _options = items;

    _listItemsValueNotifier.value = items;
  }

  void _setValue() {
    if (_options?.isNotEmpty ?? false) {
      T item = _options![_listItemFocusedPosition];
      _selectedItem = item;

      _effectiveController!.value = _selectedItem;

      if (widget.onChanged != null) {
        widget.onChanged!(_selectedItem);
      }

      setState(() {});
    }
  }

  void _clearValue() {
    T? item;
    _effectiveController!.value = item;
    _selectedItem = item;

    _listItemFocusedPosition = 0;

    if (widget.onChanged != null) {
      widget.onChanged!(_selectedItem);
    }
    _searchTextController.value = const TextEditingValue();

    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    // print("_overlayEntry : $_overlayEntry");

    _displayItem = widget.displayItemFn(_selectedItem);

    return CompositedTransformTarget(
      link: this._layerLink,
      child: GestureDetector(
        onTap: () {
          _widgetFocusNode.requestFocus();
          _toggleOverlay();
        },
        child: Focus(
          autofocus: widget.autoFocus,
          focusNode: _widgetFocusNode,
          onFocusChange: (bool focused) {
            setState(() {
              _isFocused = focused;
            });
          },
          onKey: (FocusNode focusNode, RawKeyEvent event) {
            return _onKeyPressed(event);
          },
          child: FormField(
            autovalidateMode: AutovalidateMode.always,
            initialValue: widget.searchInitialValue,
            validator: (T? str) {
              if (widget.validator != null) {
                return 'error'; //widget.validator!(_effectiveController!.value);
              }
              return null;
            },
            onSaved: (T? str) {
              if (widget.onSaved != null) {
                widget.onSaved!(_effectiveController!.value);
              }
            },
            builder: (FormFieldState<T?> state) {
              return InputDecorator(
                decoration: widget.decoration ??
                    const InputDecoration(
                      border: UnderlineInputBorder(),
                      suffixIcon: Icon(Icons.arrow_drop_down),
                    ),
                isEmpty: _isEmpty && _searchTextController.text.isEmpty,
                isFocused: _isFocused,
                child: this._overlayEntry == null
                    ? _displayItem ?? Container()
                    : EditableText(
                        controller: _searchTextController,
                        cursorColor: widget.searchCursorColor ??
                            Theme.of(context).colorScheme.primary,
                        focusNode: _searchFocusNode,
                        keyboardType: widget.searchInpuType,
                        backgroundCursorColor: Colors.transparent,
                        style: widget.searchTextStyle ??
                            Theme.of(context).textTheme.titleMedium ??
                            const TextStyle(
                              fontSize: 16,
                              color: Colors.black87,
                            ),
                        onChanged: (String str) {
                          if (_overlayEntry == null) {
                            _addOverlay();
                          }
                          _onTextChanged(str);
                          setState(() {});
                        },
                        onSubmitted: (String str) {
                          _searchTextController.value =
                              const TextEditingValue();
                          _listItemFocusedPosition = 0;
                          _setValue();
                          _removeOverlay();
                          _widgetFocusNode.nextFocus();
                        },
                        onEditingComplete: () {},
                      ),
              );
            },
          ),
        ),
      ),
    );
  }
}


example/lib/model.dart
import 'dart:convert';

class Role {
  final String name;
  final String description;
  final int id;

  Role({required this.name, required this.description, required this.id});

  factory Role.fromJson(Map<String, dynamic> json) {
    return Role(
      name: json['name'],
      description: json['desc'],
      id: json['role'],
    );
  }

  static List<Role> rolesFromJson() {
    String rolesJson = '''
    [
      {"name": "Super Admin", "desc": "Having full access rights", "role": 1},
      {"name": "Admin", "desc": "Having full access rights of a Organization", "role": 2},
      {"name": "Manager", "desc": "Having Magenent access rights of a Organization", "role": 3},
      {"name": "Technician", "desc": "Having Technician Support access rights", "role": 4},
      {"name": "Customer Support", "desc": "Having Customer Support access rights", "role": 5},
      {"name": "User", "desc": "Having End User access rights", "role": 6}
    ]
  ''';
    final List<dynamic> jsonList = jsonDecode(rolesJson);
    return jsonList.map((json) => Role.fromJson(json)).toList();
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:dropdown_plus/dropdown_plus.dart';
import 'package:example/model.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Dropdown Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Role> roles = Role.rolesFromJson();

  final DropdownEditingController<Role> _controllerRole =
      DropdownEditingController<Role>();

  final DropdownEditingController<String> _controllerGender =
      DropdownEditingController<String>();

  String? _errorMessage;

  @override
  Widget build(BuildContext context) {
    void _showModal(String value) {
      showModalBottomSheet<void>(
        context: context,
        builder: (BuildContext context) {
          return Wrap(
            children: [
              Container(
                height: 50,
                color: Colors.white,
                child: Center(
                  child: Text(value),
                ),
              )
            ],
          );
        },
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text('Dropdown Plus Demo'),
      ),
      body: SingleChildScrollView(
        child: Container(
          padding: const EdgeInsets.all(16),
          child: Column(
            children: [
              TextDropdownFormField<String>(
                options: const ["Male", "Female"],
                controller: _controllerGender,
                onChanged: (String? gender) {
                  _showModal('Gender $gender as been selected.');
                },
                validator: (String? value) {
                  return value?.isEmpty ?? false ? 'Select one' : null;
                },
                decoration: const InputDecoration(
                    border: OutlineInputBorder(),
                    suffixIcon: Icon(Icons.arrow_drop_down),
                    labelText: "Gender"),
                dropdownMaxHeight: 120,
              ),
              const SizedBox(
                height: 16,
              ),
              DropdownFormField<Role>(
                dropdownColor: Theme.of(context).colorScheme.secondary,
                emptyText: 'Role not found',
                dropdownMaxHeight: 350,
                searchCursorColor: Colors.grey[800],
                searchInpuType: TextInputType.text,
                searchInitialValue: roles[5],
                autoFocus: false,
                searchTextStyle: TextStyle(
                  color: Theme.of(context).colorScheme.secondaryContainer,
                  fontSize: 18,
                ),
                controller: _controllerRole,
                decoration: InputDecoration(
                  border: const UnderlineInputBorder(),
                  suffixIcon: const Icon(Icons.arrow_drop_down),
                  hintText: "Select a Role",
                  errorBorder: const UnderlineInputBorder(),
                  errorText: _errorMessage,
                ),
                emptyActionText: 'Create a new Role',
                onEmptyActionPressed: (String valueFromfield) async {
                  setState(() {
                    roles.add(Role(
                        name: valueFromfield,
                        description: 'Set a description for: $valueFromfield',
                        id: roles.map((Role e) => e.id).reduce(max)));
                  });
                },
                onSaved: (Role? str) {
                  _showModal('Role ${str?.name} as been selected.');
                },
                onChanged: (Role? str) {
                  _showModal('Role ${str?.name ?? ''} as been changed.');
                },
                validator: (Role? str) {
                  if (_errorMessage == null && str == null) {
                    return null;
                  }
                  _errorMessage =
                      str != null && str.name == 'User' ? 'Oh noo' : null;
      
                  return _errorMessage;
                },
                displayItemFn: (Role? item) => Text(
                  (item?.name ?? ''),
                  style: const TextStyle(fontSize: 16),
                ),
                //can make a http request
                findFn: (dynamic str) async => roles,
                selectedFn: (Role? item1, Role? item2) {
                  if (item1 != null && item2 != null) {
                    return item1.name == item2.name;
                  }
                  return false;
                },
                filterFn: (Role? item, String str) =>
                    item?.name.toLowerCase().contains(str.toLowerCase()) ?? false,
                dropdownItemFn: (Role? item, int position, bool focused,
                    bool selected, void Function() onTap) {
                  return Column(
                    children: [
                      ListTile(
                        title: Text(
                          item?.name ?? '',
                          style: TextStyle(
                            color: selected ? Colors.white : Colors.black,
                          ),
                        ),
                        subtitle: Text(
                          item?.description ?? '',
                          style: TextStyle(
                            color: selected ? Colors.white : Colors.black,
                          ),
                        ),
                        tileColor: Colors.transparent,
                        onTap: onTap,
                      ),
                      const Divider(
                        height: 3,
                        color: Colors.white60,
                      )
                    ],
                  );
                },
              ),
              const SizedBox(
                height: 50,
              ),
              ElevatedButton(
                onPressed: () {
                  _controllerRole.clear();
                  _controllerGender.clear();
                },
                child: const Text('Clear value selected'),
              )
            ],
          ),
        ),
      ),
    );
  }
}
